rex_xai/explanation.py

```diff
@@ -152,90 +204,221 @@ def _explanation(args, model_shape, prediction_func, device):
         avg_box_size,
     )
 
-    exp = Explanation(maps, prediction_func, target, data, args)  # type: ignore
+    run_stats = {
+        'total_passing': total_passing, 
+        'total_failing': total_failing, 
+        'max_depth_reached': max_depth_reached, 
+        'avg_box_size': avg_box_size
+        }
+
+    exp = Explanation(maps, prediction_func, data.target, data, args, run_stats, keep_all_maps=False)
+
+    return exp
+
+
+def analyze(exp: Explanation, data_mode: str, logging_level: int):
+    """Analyzes an Explanation.
 
-    # print(maps)
-    # for k, v in maps.items():
-    #     print(k)
-    #     print(maps.counts)
+    Analyzes the area ratio, entropy difference, insertion and deletion curves for an
+    Explanation object and prints them to ``logging.info``.
 
+    Args:
+        exp: Explanation object as returned by :py:func:`~rex_xai.explanation._explanation`
+        data_mode: Mode of the input data. Entropy difference is only calculated if ``data_mode``
+            is one of ["RGB", "L"].
+        logging_level: used to set logging level back to its original value after ensuring
+            output is printed using :py:func:`logging.info`.
+
+    Returns:
+        None
+
+    """
+    eval = Evaluation(exp)
+    rat = eval.ratio()
+    if data_mode in ("RGB", "L"):
+        be, ae = eval.entropy_loss()  # type: ignore
+        ent = be - ae
+    else:
+        ent = None
+
+    iauc, dauc = eval.insertion_deletion_curve(exp.prediction_func)
+    if logging_level < 2:
+        set_log_level(2, logger)
+    logger.info(
+        "area %f, entropy difference %f, insertion curve %f, deletion curve %f",
+        rat,
+        ent,
+        iauc,
+        dauc,
+    )
+    set_log_level(logging_level, logger)
+
+    analysis_results = {
+        "area": rat,
+        "entropy_diff": ent,
+        "insertion_curve": iauc,
+        "deletion_curve": dauc
+    }
+
+    return analysis_results
+
+
+def _explanation(
```

Comment:
This function is still doing a lot, even after the refactor. Could it be broken down into more stages? The first part, up until `exp.extract(args.strategy)`, does what I would expect this function to do, but `analyse(...)` and the collection of plotting functions seem like they could be split out from this function into their own stages.

Reply:
I think the complication with splitting out `analyze(...)` is that it's included in the timing that's happening between lines 297-305. 

I think it would also be tricky to move the plotting parts outside of the function due to how `_explanation()` is used within `explanation()` - i.e either on a single input file or inside a loop over multiple files. Same for saving to the database. To avoid having a lot of duplicated code inside `explanation`, it's helpful to have this kind of overall wrapper function that does the whole pipeline (also useful for testing all parts of the pipeline).

We could potentially group what's happening inside `_explanation()` into a few functions, though - data input, explanation, analysis, plotting? I'm just not sure if this would provide enough benefit from the simplification to justify creating additional layers/functions.